### PR TITLE
Marshal and Unmarshal methods for roomserver input api

### PR DESCRIPTION
**Pull request by NegativeMjark**

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go [hunk 1]
```diff
@@ -36,7 +40,51 @@ type InputRoomEvent struct {
 	// For example many matrix events forget to reference the m.room.create event even though it is needed for auth.
 	// (since synapse allows this to happen we have to allow it as well.)
 	AuthEventIDs []string
+	// Whether the state is supplied as a list of event IDs or whether it
+	// should be derived from the state at the previous events.
+	HasState bool
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Can't this be inferred by `len(StateEventIDs) == 0`?

Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: The state could be the empty list.

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go [hunk 2]
```diff
@@ -36,7 +40,51 @@ type InputRoomEvent struct {
...
 	// Optional list of state event IDs forming the state before this event.
 	// These state events must have already been persisted.
 	StateEventIDs []string
 }
+
+// UnmarshalJSON implements json.Unmarshaller
+func (ire *InputRoomEvent) UnmarshalJSON(data []byte) error {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Why are you implementing your own functions here? `UnmarshalJSON` seems equivalent to just using the builtin?

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go [hunk 3]
```diff
@@ -36,7 +40,51 @@ type InputRoomEvent struct {
...
+	ire.HasState = content.HasState
+	if content.Event != nil {
+		ire.Event = []byte(*content.Event)
+	}
+	return nil
+}
+
+// MarshalJSON implements json.Marshaller
+func (ire InputRoomEvent) MarshalJSON() ([]byte, error) {
+	event := json.RawMessage(ire.Event)
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Why? Please comment.

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Oh so it's actually valid to have `len(StateEventIDs) == 0` and yet that really is the room state? Comment this please, as it's pretty important and not obvious!

**Comment by Kegsay** [comment]
Other than asking for explanations/reasons, LGTM.

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Done bed4689 32af06f

File: src/github.com/matrix-org/dendrite/roomserver/api/input.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/api/input.go]: Done 32af06f

**Comment by Kegsay** [comment]
LGTM! :heart:

**Pull request merged**